### PR TITLE
bug: Allow sign and verify to use different hashes than sha256

### LINKED ISSUE #52: EC sign and verify automatically hash to SHA 256
library should accept SHA type as an optional parameter and hash to the preferred method. This bug prevents the library from being used for things like JWT, VAPID and the like.

### PR DESCRIPTION
**Pull request by jrconlin**
closes #52

### CODE REVIEW

File: pyelliptic/openssl.py [hunk 1]
```diff
@@ -100,6 +100,10 @@ def __init__(self, library):
         self.BN_bin2bn.argtypes = [ctypes.c_void_p, ctypes.c_int,
                                    ctypes.c_void_p]
 
+        self.BN_bn2dec = self._lib.BN_bn2dec
```
Comment (jcjones) [inline pyelliptic/openssl.py]: Nit: I'm not clear on why this function is being exposed in this PR. It seems fine, but maybe add a comment to the commit?

File: pyelliptic/ecc.py [hunk 2]
```diff
@@ -88,6 +90,14 @@ def __init__(self, pubkey=None, privkey=None, pubkey_x=None,
         else:
             self.privkey, self.pubkey_x, self.pubkey_y = self._generate()
 
+        _hashers = {
+            "sha256": OpenSSL.EVP_sha256(),
+            "sha384": OpenSSL.EVP_sha384(),
+            "sha512": OpenSSL.EVP_sha512(),
+        }
+        self.hashval = hasher
+        self.hasher = _hashers[hasher]
```
Comment (jcjones) [inline pyelliptic/ecc.py]: Nit: This is going to raise without an explanation if they pick an invalid hash; not sure how developer-friendly you're aiming to be.

File: pyelliptic/ecc.py [hunk 3]
```diff
@@ -414,7 +424,7 @@ def sign(self, inputb):
                 raise Exception("[OpenSSL] EC_KEY_check_key FAIL ... " + OpenSSL.get_error())
 
             OpenSSL.EVP_MD_CTX_init(md_ctx)
-            OpenSSL.EVP_DigestInit_ex(md_ctx, OpenSSL.EVP_sha256(), None)
+            OpenSSL.EVP_DigestInit_ex(md_ctx, self.hasher, None)
```
Comment (jcjones) [inline pyelliptic/ecc.py]: `EVP_DigestInit_ex` can fail (it returns 0)... that could happen if OpenSSL was built with SHA512 explicitly disabled. (That used to happen on ARMv5; probably not a huge concern...)

Still, good security coding practices say you should check the return whereever you can; that'd include the `EVP_MD_CTX_init`, too.

File: pyelliptic/ecc.py [hunk 4]
```diff
@@ -466,12 +476,10 @@ def verify(self, sig, inputb):
...
             if (OpenSSL.EC_KEY_check_key(key)) == 0:
                 raise Exception("[OpenSSL] EC_KEY_check_key FAIL ... " + OpenSSL.get_error())
-
             OpenSSL.EVP_MD_CTX_init(md_ctx)
-            OpenSSL.EVP_DigestInit_ex(md_ctx, OpenSSL.EVP_sha256(), None)
+            OpenSSL.EVP_DigestInit_ex(md_ctx, self.hasher, None)
             if (OpenSSL.EVP_DigestUpdate(md_ctx, binputb, len(inputb))) == 0:
                 raise Exception("[OpenSSL] EVP_DigestUpdate FAIL ... " + OpenSSL.get_error())
-
             OpenSSL.EVP_DigestFinal_ex(md_ctx, digest, dgst_len)
```
Comment (jcjones) [inline pyelliptic/ecc.py]: If `EVP_DigestFinal_ex` returns 0, you shouldn't use the output `digest`.

File: pyelliptic/ecc.py [hunk 5]
```diff
@@ -466,12 +476,10 @@ def verify(self, sig, inputb):
                 raise Exception("[OpenSSL] EC_KEY_set_public_key FAIL ... " + OpenSSL.get_error())
             if (OpenSSL.EC_KEY_check_key(key)) == 0:
                 raise Exception("[OpenSSL] EC_KEY_check_key FAIL ... " + OpenSSL.get_error())
-
             OpenSSL.EVP_MD_CTX_init(md_ctx)
-            OpenSSL.EVP_DigestInit_ex(md_ctx, OpenSSL.EVP_sha256(), None)
+            OpenSSL.EVP_DigestInit_ex(md_ctx, self.hasher, None)
```
Comment (jcjones) [inline pyelliptic/ecc.py]: Same as above about checking `EVP_DigestInit` and `EVP_MD_CTX_init`

File: pyelliptic/hash.py [hunk 6]
```diff
@@ -70,6 +70,18 @@ def hmac_sha256(k, m):
...
 
+def hmac_sha384(k, m):
+    """
+    Compute the key and the message with HMAC SHA384
+    """
+    key = OpenSSL.malloc(k, len(k))
+    d = OpenSSL.malloc(m, len(m))
+    md = OpenSSL.malloc(0, 64)
+    i = OpenSSL.pointer(OpenSSL.c_int(0))
+    OpenSSL.HMAC(OpenSSL.EVP_sha384(), key, len(k), d, len(m), md, i)
```
Comment (jcjones) [inline pyelliptic/hash.py]: Nit: The retval should be checked... as should the other calls to HMAC.

**Review by jcjones — COMMENTED** [review]
The code change looks OK to me; I've got some nits you should consider.

That said, in a crypto interface library, I'd strongly recommend checking return values out of OpenSSL on the important calls. For that reason, I'm not going to hit 'Approve', but as it's not my project, I'll hold off on specifically requesting changes. :)

**Comment by jrconlin** [comment]
@jcjones, thanks for the r? I'm always happy to get whatever nits or oversights folks are willing to offer, so thank you for your time. Will address and push a new version shortly.

**Review by jcjones — APPROVED** [review]
I know it's not as clean looking, but it's important to be safe with crypto code. Vulnerabilities have been built upon code not checking the return value. 👍

**Pull request closed without merging**